plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/expression/debug/Printer.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.embeddings.expression.debug
+
+import kotlin.math.max
+
+/**
+ * Helper class for pretty-printing tree views.
+ *
+ * Provides support for automatic handling of indentation and tracks line length for the user.
+ */
+class Printer(
+    /**
+     * Number of spaces to add per indent step.
+     */
+    private val indentStep: Int = 4,
+    /**
+     * Maximum line length we recommend that clients use.
+     *
+     * This is not enforced, but clients who want to produce pretty output can use this to make decisions on where to break
+     * up their output.
+     */
+    private val desiredMaxLineLength: Int = 120,
+) {
+    // _currentLine is non-null precisely when a line is currently being constructed.
+    private var _currentLine: StringBuilder? = null
+
+    // currentLine starts on a line if one isn't already being constructed.
+    private val currentLine: StringBuilder
+        get() {
+            if (_currentLine == null) {
+                _currentLine = StringBuilder(" ".repeat(indentLevel))
+            }
+            return _currentLine!!
+        }
+    private val lines = mutableListOf<String>()
+    private var indentLevel = 0
+
+    fun print(ast: TreeView) {
+        with(ast) {
+            this@Printer.printImpl()
+        }
+    }
+
+    fun print(text: String) {
+        currentLine.append(text)
+    }
+
+    fun newLine() {
+        lines.add(currentLine.toString())
+        _currentLine = null
```

Comment:
I see the meaning of this, but why aren't we using the `StringBuilder` to append a newline and indent with spaces?

```kotlin
fun newLine() {
    _currentLine.append("\n")
    _currentLine.append(" ".repeat(indentLevel))
}
```

This would allow us to have just one instance of `StringBuilder`.

Reply:
This complicates the implementation of many other points: getting the available space on the current line, adding embeds, handling the end-of-file...  Note in particular that we expect `indent` and `unindent` to be done immediately after a `newLine`, but with this implementation that would not be possible, you'd need a separate "are we working on a line?" boolean.